### ISSUE TITLE
Performance improvements of Queries

### ISSUE THREAD

**Issue by simonhac** [opennem/opennem#239]
[MBAHNTH](https://api.opennem.org.au/stats/energy/station/NEM/MBAHNTH?period=all&interval=1M) took 12.1s.

[GERMCRK](https://api.opennem.org.au/stats/energy/station/NEM/GERMCRK?period=all&interval=1M) took 9.3s.

[APPIN](https://api.opennem.org.au/stats/energy/station/NEM/APPIN?period=all&interval=1M) fails with a 404.

I really want to see these reliably responding in under 100ms. Please hook up to a monitoring service and track response times.

**Comment by nc9** [comment]
APPIN I don't think has ever had stats 

The other two are taking < 500ms for me atm - but I'll hook up pingdom to these endpoints and plot them

<img width="901" alt="image" src="https://user-images.githubusercontent.com/39009487/228522971-02888615-2148-475f-834c-cedd8a0e0041.png">

<img width="903" alt="image" src="https://user-images.githubusercontent.com/39009487/228523061-2c95020e-076b-4d3d-a672-cbe9b31234c4.png">

**Comment by simonhac** [comment]
[STANVAC](https://api.opennem.org.au/stats/energy/station/NEM/STANVAC?period=all&interval=1M) took 25.9s.

<img width="638" alt="image" src="https://user-images.githubusercontent.com/20500649/228685543-a0cd4f61-c572-4718-8b0f-e320b0be62a3.png">

**Comment by simonhac** [comment]
[LI_WY_CA](https://api.opennem.org.au/stats/energy/station/NEM/LI_WY_CA?period=all&interval=1M) took 14.5s.

<img width="835" alt="image" src="https://user-images.githubusercontent.com/20500649/228686100-46c6639c-e0b2-4836-9019-8c4e8a012aef.png">

**Comment by simonhac** [comment]
…of course they'll be much faster now they're warmed up.

**Comment by nc9** [comment]
* profile where it is slow 
* create `at_facility_monthly` for monthly aggregates

**Comment by nc9** [comment]
`at_facility_daily` isn't a timescale managed hypertable, i'm making it one on dev:

https://docs.timescale.com/use-timescale/latest/hypertables/about-hypertables/

made a separate branch-build to test performance

**Comment by nc9** [comment]
related #196

**Comment by nc9** [comment]
1 hour lump this morning of *horrible* performance and 100% cpu - investigating now

**Comment by gmmspi** [comment]
Will publish fixes to prod and check Pingdom again on Monday to verify

**Comment by nc9** [comment]
![image](https://user-images.githubusercontent.com/39009487/236710282-bb847b2f-7271-4e4d-98fc-7eacb558a627.png)

**Comment by nc9** [comment]
this will improve further with #88

**Comment by gmmspi** [comment]
Significant speed improvements noted across all examples in OG issue

**Comment by morganchristiansson** [comment]
Hello. With stats API now redirecting to 7d.json how can I get 5m resolution historical data? Changed in this commit: https://github.com/opennem/opennem/commit/e3dd9ae3e94dabaf625a1f4bc9541f3cd35c8ea0

From searching this repo I managed to find these static json for per-week 5m data - but they are missing rooftop solar data?
https://data.opennem.org.au/v3/stats/historic/weekly/NEM/NSW1/year/2023/week/24.json

What is official way to obtain historical 5m data incl rooftop solar? Thanks.

**Comment by morganchristiansson** [comment]
Have you considered pre-calculating this query using a [continous aggregate](https://docs.timescale.com/use-timescale/latest/continuous-aggregates/about-continuous-aggregates/)?
There's 500+ scada facilities making this query very slow.

This should allow you to restore the [/stats/power/network/fueltech/{network_code}/{network_region_code}](https://api.opennem.org.au/docs#tag/Stats/operation/Power_Network_Region_by_Fueltech_stats_power_network_fueltech__network_code___network_region_code__get) API endpoint

https://github.com/opennem/opennem/blob/d0dfcff9d7531243db3352cef3da4a7dfb82f625/opennem/api/export/queries.py#L373-L394

**Comment by nc9** [comment]
Hi Christian - that is exactly what we're doing in V4 - when we first started using continuous aggregates with timescaledb ~2 years ago we found that they weren't flexible enough and the updates were too slow (not keeping the data live) so we removed them all in favour of our aggregate tables for energy and just running the power queries live.

We just can't have a fully open API anymore - we're a non-profit, already under-resourced on the dev side (I do both backend + infra and am not fulltime) but with v4 and API keys we're hoping to take some control back, allow full access to the entire data history, and compensate our costs with pro API accounts for commercial users.

**Comment by nc9** [comment]
Oh - and we upgraded to SqlAlchemy 2, all async, param'd `text` queries etc. etc.

**Comment by morganchristiansson** [comment]
Yeah I understand. You could also upgrade the aggregated table from python using `INSERT INTO SELECT ... ON CONFLICT` as you insert new data into the facility_scada table, preferable if the DB does it for you, but if that's not working well enough...

A plain select query without aggregation should be fast enough.

**Comment by nc9** [comment]
I'm re-opening this and tagging for v4 because there are improvements we can make for the v4 API

After some reading earlier found timescale now supports real-time continuous aggregates:

https://docs.timescale.com/use-timescale/latest/continuous-aggregates/real-time-aggregates/

@morganchristiansson I like that suggestion. It's how we build our aggregate tables now, but we can do similar for all API endpoints and more recent power data

FYI when I wrote those first aggregates and benchmarked different approaches, this worked best for getting bulk data in:

https://github.com/opennem/opennem/blob/38ea69234dcf3fc6b690d0497d50d3c8b57cf8d8/opennem/db/bulk_insert_csv.py

(that file could use a cleanup) .. ON CONFLICT DO UPDATE's are literered throughout the codebase (see [this search](https://github.com/search?q=repo%3Aopennem%2Fopennem%20on_conflict&type=code))

If you're curious our aggregate tables are all in the ORM starting here:

https://github.com/opennem/opennem/blob/master/opennem/db/models/opennem.py#L847

and prefixed with `at_` - the facility history pages all query out at <100ms 

they're all running out of `opennem.aggregates` 

There are two issues here:

1. Historic data and the current `at_` tables - make sure the API is querying them where it can and expose all of that day 0 data in the API. This is pretty much a manual approach materializing the view.
2. Figure out an approach for 'live' data that doesn't involve that ugly code block in the API you pointed out, where it sounds like the timescale realtime aggregates may be worth investigating again

**Comment by morganchristiansson** [comment]
I can also say that just dumping all data as JSON/CSV files is a _genius_ idea that scales to infinity with a CDN at low cost and deletes most of the problems. Don't need an API! We overcomplicate the solutions.
Obviously AEMO does this. Also Canada IESO http://reports.ieso.ca/public/ and California CAISO (though they don't have directory listings) http://www.caiso.com/outlook/SP/History/%Y%m%d/{fuelsource,netdemand}.csv

You could just expand your existing dumps to cover the full history.
That said I think current API will perform very well with simple and performant SELECT queries after aggregations.

An API rate limit might also be relevant to force aggressive API clients to back off. Can be difficult to get right with CDN and multiple web servers. But could rate limit on server load also.

Sounds like you have the right ideas. <100ms still sounds high. Happy to bounce ideas.

---

Re bulk copy my INSERT INTO SELECT hasn't had much issues yet, though it sometimes takes 10s of minutes to run.

Interesting that COPY would be faster, I only use this method between databases in same server right now. I am using ruby tool [pgsync](https://github.com/ankane/pgsync) which uses COPY and STDIN. It's not without issues for very large copies.

I haven't had a chance to look at it yet but https://github.com/timescale/timescaledb-parallel-copy seems like a better tool. Thought it also seems to use COPY .. STDIN .. Maybe it just does batches in parallel optimized for timescaledb chunks, pgsync can only do chunk when there is an id column.

**Comment by nc9** [comment]
> You could just expand your existing dumps to cover the full history.

It's there - we just haven't promoted it yet because we want to validate all the data. If you look at some of the open issues, you'll see we have data gaps. We also completely changed how we calculate flows and flow emissions, so that history needs to be rewritten, see:

https://github.com/opennem/opennem/issues/65#issuecomment-1216055848

> An API rate limit might also be relevant to force aggressive API clients to back off.

The v4 API will all be behind API keys with rate limiting (we're using unkey.dev). The current API will steadily be deprecated over months with plenty of warning. It's still being _really_ abused at the moment.

We have this form for API key signup - we'll send invites out steadily:

https://github.com/opennem/opennem/discussions/243

> I haven't had a chance to look at it yet but https://github.com/timescale/timescaledb-parallel-copy seems like a better tool. 

Oh thank you - this is brilliant. I have a (poorer and domain-specific) version of this in python by breaking the data up into chunks and using multiprocessing, but it took me ~5 min to understand that code and how he's using channels. Interesting that his default chunk size of 5000 is pretty close to the default chunk size I came to after some trial-and-error (I think I settled on 3600 or some weird number like that)

OpenNEM v4 is all async with sqlalchemy2 + asyncpg and pydantic2 for parsing the external data - I'd be curious how performance compares

**Comment by nc9** [comment]
Here's one answer:

https://magic.io/blog/asyncpg-1m-rows-from-postgres-to-python/

😮

**Comment by nc9** [comment]
> Sounds like you have the right ideas. <100ms still sounds high. Happy to bounce ideas.

Forgot to ACK but thank you - will def keep this thread open with updates. I'm going to start a branch ~soon (which at current schedule means likely late this week) and test the new realtime continuous aggregates on _exactly_ that query you suggested with the live power output.

I'll post results here.
